fiaas_deploy_daemon/deployer/kubernetes/ingress.py

```diff
@@ -51,38 +53,72 @@ def delete(self, app_spec):
+        unannotated_ingress = AnnotatedIngress(name=app_spec.name, ingress_items=[], annotations={})
+        ingresses_by_annotations = [unannotated_ingress]
+        for ingress_item in app_spec.ingresses:
+            if ingress_item.annotations:
+                next_name = "{}-{}".format(app_spec.name, len(ingresses_by_annotations))
+                annotated_ingresses = AnnotatedIngress(name=next_name, ingress_items=[ingress_item], annotations=ingress_item.annotations)
+                ingresses_by_annotations.append(annotated_ingresses)
+            else:
+                unannotated_ingress.ingress_items.append(ingress_item)
+
+        for annotated_ingress in ingresses_by_annotations:
+            if len(annotated_ingress.ingress_items) == 0:
+                continue
+
+            self._create_ingress(app_spec, annotated_ingress, custom_labels)
+
+        self._delete_unused(custom_labels)
+
+    @retry_on_upsert_conflict
+    def _create_ingress(self, app_spec, annotated_ingress, labels):
+        default_annotations = {
+            u"fiaas/expose": u"true" if _has_explicitly_set_host(annotated_ingress.ingress_items) else u"false"
         }
+        annotations = merge_dicts(annotated_ingress.annotations, app_spec.annotations.ingress, default_annotations)
 
-        custom_labels = merge_dicts(app_spec.labels.ingress, labels)
-        custom_annotations = merge_dicts(app_spec.annotations.ingress, annotations)
-        metadata = ObjectMeta(name=app_spec.name, namespace=app_spec.namespace, labels=custom_labels,
-                              annotations=custom_annotations)
+        metadata = ObjectMeta(name=app_spec.name, namespace=app_spec.namespace, labels=labels,
+                                annotations=annotations)
 
         per_host_ingress_rules = [
             IngressRule(host=self._apply_host_rewrite_rules(ingress_item.host),
                         http=self._make_http_ingress_rule_value(app_spec, ingress_item.pathmappings))
-            for ingress_item in app_spec.ingresses
+            for ingress_item in annotated_ingress.ingress_items
             if ingress_item.host is not None
         ]
-        default_host_ingress_rules = self._create_default_host_ingress_rules(app_spec)
+        if annotated_ingress.annotations:
+            default_host_ingress_rules = []
+        else:
+            default_host_ingress_rules = self._create_default_host_ingress_rules(app_spec)
+
+        print per_host_ingress_rules + default_host_ingress_rules
 
         ingress_spec = IngressSpec(rules=per_host_ingress_rules + default_host_ingress_rules)
 
         ingress = Ingress.get_or_create(metadata=metadata, spec=ingress_spec)
         self._ingress_tls.apply(ingress, app_spec, self._get_hosts(app_spec))
         ingress.save()
 
+    def _delete_unused(self, labels):
+        filter_labels = {
+            "app": Equality(labels["app"]),
+            "fiaas/deployment_id": Inequality(labels["fiaas/deployment_id"])
```

Comment:
I think this need another filter: "fiaas/deployment_id": Exists()

Otherwise, it will match all ingresses from that does not have a fiaas/deployment_id (e.g. custom ingresses for that app created off-band). This feature should make these extra ingresses not necessary, but still seems like better fall on the safe side?

Reply:
Ah, true. But passing a map with duplicate keys isn't possible :/

Reply:
Maybe like this? Not sure if it's the tidiest API: https://github.com/fiaas/k8s/pull/80

Reply:
I think that solution is good. I thought about a map where the values may be lists, but I think the list of tuples better represents what the k8s API is.